File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/fragment/task/SearchFragment.java
package hcm.ditagis.com.cholon.qlsc.fragment.task;

import android.annotation.SuppressLint;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.v4.app.Fragment;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.esri.arcgisruntime.data.CodedValue;
import com.esri.arcgisruntime.data.CodedValueDomain;
import com.esri.arcgisruntime.data.Domain;
import com.esri.arcgisruntime.data.Feature;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.List;

import hcm.ditagis.com.cholon.qlsc.ListTaskActivity;
import hcm.ditagis.com.cholon.qlsc.R;
import hcm.ditagis.com.cholon.qlsc.async.QueryFeatureAsync;
import hcm.ditagis.com.cholon.qlsc.entities.DApplication;
import hcm.ditagis.com.cholon.qlsc.utities.Constant;


@SuppressLint("ValidFragment")
public class SearchFragment extends Fragment {
    private View mRootView;
    private ListTaskActivity mActivity;
    private EditText mEtxtAddress;
    private Spinner mSpinTrangThai;
    private TextView mTxtThoiGian;
    private Button mBtnSearch;
    private TextView mTxtKetQua;
    private LinearLayout mLLayoutKetQua;
    private DApplication mApplication;
    private List<CodedValue> mCodeValues;
    private List<Feature> mFeaturesResult;

    @SuppressLint("ValidFragment")
    public SearchFragment(ListTaskActivity activity, final LayoutInflater inflater) {
        mRootView = inflater.inflate(R.layout.fragment_list_task_search, null);
        this.mActivity = activity;
        mApplication = (DApplication) activity.getApplication();
        init();
    }

    private void init() {
        mEtxtAddress = mRootView.findViewById(R.id.etxt_list_task_search_address);
        mSpinTrangThai = mRootView.findViewById(R.id.spin_list_task_search_trang_thai);
        mTxtThoiGian = mRootView.findViewById(R.id.txt_list_task_search_thoi_gian);
        mBtnSearch = mRootView.findViewById(R.id.btn_list_task_search);
        mLLayoutKetQua = mRootView.findViewById(R.id.llayout_list_task_search_ket_qua);
        mTxtKetQua = mRootView.findViewById(R.id.txt_list_task_ket_qua);

        mBtnSearch.setOnClickListener(this::onClick);
        mTxtThoiGian.setOnClickListener(this::onClick);
        initSpinTrangThai();
    }

    private void initSpinTrangThai() {
        Domain domain = mApplication.getDFeatureLayer().getLayer().getFeatureTable().getField(Constant.FieldSuCo.TRANG_THAI).getDomain();
        if (domain != null) {
            mCodeValues = ((CodedValueDomain) domain).getCodedValues();
            if (mCodeValues != null) {
                List<String> codes = new ArrayList<>();
                codes.add("Tất cả");
                for (CodedValue codedValue : mCodeValues)
                    if (!Constant.DEFINITION_HIDE_COMPLETE.contains(codedValue.getCode().toString()))
                        codes.add(codedValue.getName());
                ArrayAdapter<String> adapter = new ArrayAdapter<>(mRootView.getContext(), android.R.layout.simple_list_item_1, codes);
                mSpinTrangThai.setAdapter(adapter);
            }
        }
    }

    private void showDateTimePicker() {
        final View dialogView = View.inflate(mRootView.getContext(), R.layout.date_time_picker, null);
        final android.app.AlertDialog alertDialog = new android.app.AlertDialog.Builder(mRootView.getContext()).create();
        dialogView.findViewById(R.id.date_time_set).setOnClickListener(view -> {
            DatePicker datePicker = dialogView.findViewById(R.id.date_picker);
            Calendar calendar = new GregorianCalendar(datePicker.getYear(), datePicker.getMonth(), datePicker.getDayOfMonth());
            String displaytime = (String) DateFormat.format((Constant.DateFormat.DATE_FORMAT_STRING), calendar.getTime());
            @SuppressLint("SimpleDateFormat") SimpleDateFormat dateFormatGmt = Constant.DateFormat.DATE_FORMAT_YEAR_FIRST;
//            dateFormatGmt.setTimeZone(TimeZone.getTimeZone("GMT"));
            mTxtThoiGian.setText(displaytime);
            alertDialog.dismiss();
        });
        alertDialog.setView(dialogView);
        alertDialog.show();

    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private void traCuu() {
        mLLayoutKetQua.removeAllViews();
        short trangThai = -1;
        if (mCodeValues != null)
            for (CodedValue codedValue : mCodeValues) {
                if (codedValue.getName().equals(mSpinTrangThai.getSelectedItem().toString())) {
                    trangThai = Short.parseShort(codedValue.getCode().toString());
                }
            }
        new QueryFeatureAsync(mActivity, trangThai,
                mEtxtAddress.getText().toString(),
                mTxtThoiGian.getText().toString(), output -> {
            if (output != null && output.size() > 0) {
                mFeaturesResult = output;
                List<View> views = HandlingSearchHasDone.handleFromFeatures(mActivity, mRootView.getContext(), output);
                for (View view : views) {
                    TextView txtID = view.findViewById(R.id.txt_top);
                    view.setOnClickListener(v -> {
                        mActivity.itemClick(txtID.getText().toString());
                    });
                    mLLayoutKetQua.addView(view);
                }
                mTxtKetQua.setVisibility(View.VISIBLE);
                mTxtKetQua.setText(String.format("Kết quả tra cứu: %d sự cố", mLLayoutKetQua.getChildCount()));
            } else {
                mTxtKetQua.setVisibility(View.INVISIBLE);
                Toast.makeText(mRootView.getContext(), "Không có kết quả", Toast.LENGTH_SHORT).show();
            }
        }).execute();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return mRootView;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_list_task_search:
                traCuu();
                break;
            case R.id.txt_list_task_search_thoi_gian:
                showDateTimePicker();
                break;
        }
    }
}

File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/fragment/task/SearchFragment.kt
package hcm.ditagis.com.cholon.qlsc.fragment.task

import android.annotation.SuppressLint
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.esri.arcgisruntime.data.CodedValue
import com.esri.arcgisruntime.data.CodedValueDomain
import com.esri.arcgisruntime.data.Feature
import hcm.ditagis.com.cholon.qlsc.ListTaskActivity
import hcm.ditagis.com.cholon.qlsc.R
import hcm.ditagis.com.cholon.qlsc.async.QueryFeatureAsync
import hcm.ditagis.com.cholon.qlsc.entities.DApplication
import hcm.ditagis.com.cholon.qlsc.utities.Constant
import java.util.*


@SuppressLint("ValidFragment")
class SearchFragment @SuppressLint("ValidFragment")
constructor(private val mActivity: ListTaskActivity, inflater: LayoutInflater) : Fragment() {
    private val mRootView: View = inflater.inflate(R.layout.fragment_list_task_search, null)
    private var mEtxtAddress: EditText? = null
    private var mSpinTrangThai: Spinner? = null
    private var mTxtThoiGian: TextView? = null
    private var mBtnSearch: Button? = null
    private var mTxtKetQua: TextView? = null
    private var mLLayoutKetQua: LinearLayout? = null
    private val mApplication: DApplication
    private var mCodeValues: List<CodedValue>? = null
    private var mFeaturesResult: List<Feature>? = null

    init {
        mApplication = mActivity.application as DApplication
        init()
    }

    private fun init() {
        mEtxtAddress = mRootView.findViewById(R.id.etxt_list_task_search_address)
        mSpinTrangThai = mRootView.findViewById(R.id.spin_list_task_search_trang_thai)
        mTxtThoiGian = mRootView.findViewById(R.id.txt_list_task_search_thoi_gian)
        mBtnSearch = mRootView.findViewById(R.id.btn_list_task_search)
        mLLayoutKetQua = mRootView.findViewById(R.id.llayout_list_task_search_ket_qua)
        mTxtKetQua = mRootView.findViewById(R.id.txt_list_task_ket_qua)

        mBtnSearch!!.setOnClickListener{traCuu()}
        mTxtThoiGian!!.setOnClickListener {showDateTimePicker() }
        initSpinTrangThai()
    }

    private fun initSpinTrangThai() {
        val domain = mApplication.dFeatureLayer!!.layer.featureTable.getField(Constant.FieldSuCo.TRANG_THAI).domain
        if (domain != null) {
            mCodeValues = (domain as CodedValueDomain).codedValues
            if (mCodeValues != null) {
                val codes = ArrayList<String>()
                codes.add("Tất cả")
                for (codedValue in mCodeValues!!)
                    if (!Constant.DEFINITION_HIDE_COMPLETE.contains(codedValue.code.toString()))
                        codes.add(codedValue.name)
                val adapter = ArrayAdapter(mRootView.context, android.R.layout.simple_list_item_1, codes)
                mSpinTrangThai!!.adapter = adapter
            }
        }
    }

    private fun showDateTimePicker() {
        val dialogView = View.inflate(mRootView.context, R.layout.date_time_picker, null)
        val alertDialog = android.app.AlertDialog.Builder(mRootView.context).create()
        dialogView.findViewById<View>(R.id.date_time_set).setOnClickListener { view ->
            val datePicker = dialogView.findViewById<DatePicker>(R.id.date_picker)
            val calendar = GregorianCalendar(datePicker.year, datePicker.month, datePicker.dayOfMonth)
            val displaytime = DateFormat.format(Constant.DateFormat.DATE_FORMAT_STRING, calendar.time) as String
            @SuppressLint("SimpleDateFormat") val dateFormatGmt = Constant.DateFormat.DATE_FORMAT_YEAR_FIRST
            //            dateFormatGmt.setTimeZone(TimeZone.getTimeZone("GMT"));
            mTxtThoiGian!!.text = displaytime
            alertDialog.dismiss()
        }
        alertDialog.setView(dialogView)
        alertDialog.show()

    }


    private fun traCuu() {
        mLLayoutKetQua!!.removeAllViews()
        var trangThai: Short = -1
        if (mCodeValues != null)
            for (codedValue in mCodeValues!!) {
                if (codedValue.name == mSpinTrangThai!!.selectedItem.toString()) {
                    trangThai = java.lang.Short.parseShort(codedValue.code.toString())
                }
            }
        QueryFeatureAsync(mActivity, trangThai.toInt(),
                mEtxtAddress!!.text.toString(),
                mTxtThoiGian!!.text.toString(), object: QueryFeatureAsync.AsyncResponse {
            override fun processFinish(output: List<Feature>?) {
                if (output != null && output.isNotEmpty()) {
                    mFeaturesResult = output
                    val views = HandlingSearchHasDone.handleFromFeatures(mActivity, mRootView.context, output)
                    for (view in views) {
                        val txtID = view.findViewById<TextView>(R.id.txt_top)
                        view.setOnClickListener { v -> mActivity.itemClick(txtID.text.toString()) }
                        mLLayoutKetQua!!.addView(view)
                    }
                    mTxtKetQua!!.visibility = View.VISIBLE
                    mTxtKetQua!!.text = String.format("Kết quả tra cứu: %d sự cố", mLLayoutKetQua!!.childCount)
                } else {
                    mTxtKetQua!!.visibility = View.INVISIBLE
                    Toast.makeText(mRootView.context, "Không có kết quả", Toast.LENGTH_SHORT).show()
                }
            }
        }).execute()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return mRootView
    }


}
